cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,99 @@
+    temp.release(stream);
+}
+```
+## Using thrust
+To ensure that thrust algorithms allocate temporary memory via the provided device memory allocator the `ML::thrustAllocatorAdapter` available in `allocatorAdapter.hpp` should be used with the `thrust::cuda::par` execution policy:
+```cpp
+void foo( ML::cumlHandle* handle, .. )
+{
+    ML::thrustAllocatorAdapter alloc( handle->getDeviceAllocator(), handle->getStream() );
+    auto execution_policy = thrust::cuda::par(alloc).on(handle->getStream());
+    thrust::for_each(execution_policy, ... );
+}
+```
+# Asynchronous operations and stream ordering
+All ML algorithms should be as asynchronous as possible avoiding the use of the default stream (aka as NULL or `0` stream). If an implementation only requires a single CUDA Stream the stream from `ML::cumlHandle` should be used:
+```cpp
+void foo( ML::cumlHandle* handle, ...)
+{
+    cudaStream_t stream = handle->getStream();
+}
+```
+In case multiple streams are needed, e.g. to manage a pipeline, the internal streams available in `ML::cumlHandle_impl` should be used (see [CUDA Resources](# CUDA Resources)). If multiple streams are used all operations still need to be ordered according to `ML::cumlHandle::getStream()`, i.e. before any operation in any of the internal CUDA streams is started all previous work in `ML::cumlHandle: getStream()` needs to have completed and any work enqueued in `ML::cumlHandle::getStream()` after an cuML function returns should not start before all work enqueued in the internal stream has complete. This can be ensured by introducing inter stream dependencies with CUDA events and `cudaStreamWaitEvent`. For convenience  the header `cumlHandle.hpp` provide the class `ML::detail::streamSyncer` which lets all `ML::cumlHandle_impl` internal CUDA streams wait on `ML::cumlHandle::getStream()` in its constructor and in its destructor and lets `ML::cumlHandle::getStream()` wait on all work enqueued in the `ML::cumlHandle_impl` internal CUDA streams. E.g. a usage like this:
+```cpp
+void cumlAlgo( ML::cumlHandle* handle, ...)
+{
+    ML::detail::streamSyncer _(handle);
+}
+```
+ensures the stream ordering behavior described above.
+
+## Using thrust
+To ensure that thrust algorithms are executed in the intended stream the `thrust::cuda::par` execution policy should be used (described in the section Device and Host memory allocations/Using thrust).
+
+# CUDA Resources
+Implementations of ML algorithms should not create reusable resources themselves. Instead they should use the existing one in `ML::cumlHandle_impl `. This allows to avoid constant creating and recreation of reusable resources such as CUDA streams, CUDA events or library handles. Please file a feature request in case a resource handle is missing in `ML::cumlHandle_impl `.
+The resources can be obtained like this
+```cpp
+void foo( ML::cumlHandle* handle, ...)
+{
+    cublasHandle_t cublasHandle = handle->getImpl()->getCublasHandle();
```

Comment:
@jirikraus all of our cu*Handle objects typically support a `setStream` method and IIRC they all have default stream initialized in them? Maybe we should ask devs to prohibit from calling `setStream` on these handles?

Reply:
Or maybe update our `cumlHandle_impl::setStream` to in turn also set streams for these handles appropriately?

Reply:
That is a good point. `setStream` is not really meant for cuML developers. It is meant for cuML users. Therefore I think it is better to pass around `const ML::cumlHandle&` in cuML which would prohibit changing the user provided stream. This also ties in with a comment from @jrhemstad regarding how to pass around the handle.

Reply:
Sorry Jiri, if my previous comments confused you. I meant the following kind of usage pattern. In other words, should we allow developers to perform cublasSetStream on the internal cublas handle, explicitly, for example?
```
cublasHandle_t cublasHandle = handle->getImpl()->getCublasHandle();
cudaStream_t someStreamPickedByMe;
cublasSetStream(cublasHandle, someStreamPickedByMe);
```

Reply:
On one hand, it sounds like a valid use-case if we want to allow devs to use internal streams and schedule cublas/cusolver calls on these streams.

On the other hand, these `cublasSetStream` kind of calls are persistent, easily causing false dependencies when the same cublasHandle gets called somewhere else in the pipeline!

Reply:
I also thought a bit more about this and one possibility would be that we drop the assumption that e.g. the `cublasHandle` managed by handle has any stream associated with it. Instead one should assume that it is required to set the stream before every cuBLAS call. For cuML and ml-prims developers I think everybody should use the wrapper is `cublas_wrappers.h` and we add a stream argument to the interface of those wrappers. E.g. for gemm
```cpp
template <typename T>
cublasStatus_t cublasgemm(cublasHandle_t handle, cublasOperation_t transA,
                          cublasOperation_t transB, int m, int n, int k,
                          const T *alfa, const T *A, int lda, const T *B,
                          int ldb, const T *beta, T *C, int ldc, cudaStream_t stream);
```
which would have the following specialization for dgemm
```cpp
template <>
inline cublasStatus_t
  cublasgemm(cublasHandle_t handle, cublasOperation_t transA,
             cublasOperation_t transB, int m, int n, int k, const double *alfa,
             const double *A, int lda, const double *B, int ldb,
             const double *beta, double *C, int ldc, cudaStream_t stream) {
  cublasStatus_t status = cublasSetStream(handle, stream);
  if ( CUBLAS_STATUS_SUCCESS != status )
    return status;
  return cublasDgemm(handle, transA, transB, m, n, k, alfa, A, lda, B, ldb,
                     beta, C, ldc);
}
```

Rational for this idea is that changing streams is s check operation and after an operation is en queued it is fine to change the stream. Only issue I see with this is potential races in a multi threaded enviroment, e.g. 
Thread A: sets stream to stream A on handle
Thread B: sets stream to stream B on handle
Thread B: launches Dgemm on handle in stream B
Thread B: launches Dgemm on handle in stream B (ups!)
We could either manage that by providing a cublasHandle_t wrapper containing a lock in addition to the handle that could be used in the cuBLAS wrappers. Or simply define a cuML, ml-prims developer rule that the wrapper are not thread safe and that one handle should be created per thread? The latter one would then require the `cumlHandle_t` to mange multiple cublasHandle's or that we disallow multi threading for cuML, ml-prims implementation. Considering the cuBLAS documentation having one handle per thread is propably what we should prefer:

[2.1.3. Thread Safety](https://docs.nvidia.com/cuda/cublas/index.html#thread-safety2)
The library is thread safe and its functions can be called from multiple host threads, even with the same handle. When multiple threads share the same handle, extreme care needs to be taken when the handle configuration is changed because that change will affect potentially subsequent CUBLAS calls in all threads. It is even more true for the destruction of the handle. So it is not recommended that multiple thread share the same CUBLAS handle.

What do you think?

Reply:
My opinion (might be based, because I like the OPG approach!) is that we should disallow the usage of same cumlHandle across multiple threads.

Your cublas_wrappers idea is a much better one.

Implementation details:
If we allow `stream` as a default option in the above `cublasgemm` method you defined above, then I think this change can be made within the branch-0.6 timeframe itself. Else, it'll cause huge amount of changes across existing ml-prims and cuML code (not to mention the rework needed for ongoing PR's!).

Reply:
Have filed issue #256 as our first step towards updating these *_wrapper header files. We shall update the existing methods to expose a stream as a default argument.

Will soon file another issue to remove the default-ness of stream argument across **all** ml-prims methods. We should not be launching work on default streams inside cuML or ml-prims!

Reply:
Agree. It makes sense that the implementation of cuML and ml-prims is single threaded in this case we won't have the issues I described. Just added two sentences in the Thread safety section of `DEVELOPER_GUIDE.md` to document it and put it up for discussion.